src/lib/structures/Josh.ts

```diff
@@ -0,0 +1,1813 @@
+   * @param hook The hook to run for the payload.
+   * @returns The {@link Josh} instance.
+   *
+   * @example
+   * ```typescript
+   * josh.use({ name: 'my-middleware' }, (payload) => payload);
+   * ```
+   */
+  public use<P extends Payload>(options: Josh.UseMiddlewareOptions, hook: (payload: P) => Awaitable<P>): this;
+
+  /**
+   * Adds a middleware by providing a {@link Middleware} instance.
+   * @since 2.0.0
+   * @param instance The middleware instance.
+   * @returns The {@link Josh} instance.
+   *
+   * @example
+   * ```typescript
+   * josh.use(new MyMiddleware());
+   * ```
+   */
+  public use(instance: Middleware<StoredValue>): this;
+  public use<P extends Payload>(optionsOrInstance: Josh.UseMiddlewareOptions | Middleware<StoredValue>, hook?: (payload: P) => Awaitable<P>): this {
+    if (optionsOrInstance instanceof Middleware) this.middlewares.set(optionsOrInstance.name, optionsOrInstance);
+    else {
+      if (hook === undefined)
+        throw new JoshError({
+          identifier: Josh.Identifiers.UseMiddlewareHookNotFound,
+          message: 'The "hook" parameter for middleware was not found.'
+        });
+
+      const { name, position, trigger, method } = optionsOrInstance;
+      const options: Middleware.Options = { name, position, conditions: { pre: [], post: [] } };
+      const middleware = this.middlewares.get(options.name) ?? new Middleware<StoredValue>(options).init(this.middlewares);
+
+      if (trigger !== undefined && method !== undefined) options.conditions[trigger === Trigger.PreProvider ? 'pre' : 'post'].push(method);
+
+      Object.defineProperty(middleware, method === undefined ? 'run' : method, { value: hook });
+
+      this.middlewares.set(middleware.name, middleware);
+    }
+
+    return this;
+  }
+
+  /**
+   * Generate an automatic key. Generally an integer incremented by `1`, but depends on provider.
+   * @since 2.0.0
+   * @returns The newly generated automatic key.
+   *
+   * @example
+   * ```typescript
+   * const key = await josh.autoKey();
+   *
+   * await josh.set(key, 'value');
+   * ```
+   */
+  public async autoKey(): Promise<string> {
+    let payload: AutoKeyPayload = { method: Method.AutoKey, trigger: Trigger.PreProvider, data: '' };
+
+    for (const middleware of this.middlewares.array()) await middleware.run(payload);
+    for (const middleware of this.middlewares.getPreMiddlewares(Method.AutoKey)) payload = await middleware[Method.AutoKey](payload);
+
+    payload = await this.provider[Method.AutoKey](payload);
+    payload.trigger = Trigger.PostProvider;
+
+    if (payload.error) throw payload.error;
+
+    for (const middleware of this.middlewares.array()) await middleware.run(payload);
+    for (const middleware of this.middlewares.getPostMiddlewares(Method.AutoKey)) payload = await middleware[Method.AutoKey](payload);
+
+    return payload.data;
+  }
+
+  /**
+   * Clears all stored values from the provider.
```

Comment:
A stronger wording here might be important. **Deletes** all values, and maybe even a warning about this deletes everything and can't be undone.